### DISCUSSION TITLE
Examples of YOLO augmentations

### DISCUSSION THREAD

**Discussion by NS-Nik** [ultralytics/ultralytics#15637] [Q&A]
Hello! 

I am using the YOLOv8 in my research and I need to visualize how augmentation works for my dataset. 
I understand that I can look into the train_batch images. But thats not what I am looking for. 

Is it possible to generate a set of examples of YOLO augmentations for the specific image?

**Comment by glenn-jocher** [comment]
@NS-Nik hello! To visualize YOLO augmentations for a specific image, you can use the `albumentations` library directly in your code. This allows you to apply the same augmentations used in YOLOv8. For more details on the specific augmentations supported, please refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by NS-Nik** [reply]
Thank you for the answer! 
I checked the  albumentations library and tried the following example:

```
>>> from ultralytics.data.dataset import YOLODataset
>>> dataset = YOLODataset(img_path='path/to/images', imgsz=640)
>>> hyp = {'mosaic': 1.0, 'copy_paste': 0.5, 'degrees': 10.0, 'translate': 0.2, 'scale': 0.9}
>>> transforms = v8_transforms(dataset, imgsz=640, hyp=hyp)
>>> augmented_data = transforms(dataset[0])

```

However, I am getting an error on the YOLODataset stage. 
It is not very clear from the documentation,  whether _img_path_ refers to the path the folder with images, the parent directory with the images and labels folders or to the .yaml  config file. 
Either way, after running this line of the code I get the same error:

```
AttributeError                            Traceback (most recent call last)
Cell In[49], line 1
----> 1 YOLODataset(img_path='/Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/arctic/data/train/images',imgsz=640)

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/dataset.py:41, in YOLODataset.__init__(self, data, use_segments, use_keypoints, *args, **kwargs)
     39 self.data = data
     40 assert not (self.use_segments and self.use_keypoints), 'Can not use both segments and keypoints.'
---> 41 super().__init__(*args, **kwargs)

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/base.py:97, in BaseDataset.__init__(self, img_path, imgsz, cache, augment, hyp, prefix, rect, batch_size, stride, pad, single_cls, classes, fraction)
     94     self.cache_images(cache)
     96 # Transforms
---> 97 self.transforms = self.build_transforms(hyp=hyp)

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/dataset.py:142, in YOLODataset.build_transforms(self, hyp)
    140     hyp.mosaic = hyp.mosaic if self.augment and not self.rect else 0.0
    141     hyp.mixup = hyp.mixup if self.augment and not self.rect else 0.0
--> 142     transforms = v8_transforms(self, self.imgsz, hyp)
    143 else:
    144     transforms = Compose([LetterBox(new_shape=(self.imgsz, self.imgsz), scaleup=False)])

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/augment.py:931, in v8_transforms(dataset, imgsz, hyp, stretch)
    919 """Convert images to a size suitable for YOLOv8 training."""
    920 pre_transform = Compose([
    921     Mosaic(dataset, imgsz=imgsz, p=hyp.mosaic),
    922     CopyPaste(p=hyp.copy_paste),
   (...)
    929         pre_transform=None if stretch else LetterBox(new_shape=(imgsz, imgsz)),
    930     )])
--> 931 flip_idx = dataset.data.get('flip_idx', [])  # for keypoints augmentation
    932 if dataset.use_keypoints:
    933     kpt_shape = dataset.data.get('kpt_shape', None)

AttributeError: 'NoneType' object has no attribute 'get'
```

**Reply by glenn-jocher** [reply]
@NS-Nik it looks like the `YOLODataset` initialization is causing the issue. The `data` parameter should point to your dataset configuration `.yaml` file, not the image directory. Please update your code to use the path to your `.yaml` file and try again. If the issue persists, ensure you're using the latest version of the Ultralytics package.

**Reply by NS-Nik** [reply]
@glenn-jocher Thank you for your work and help! 

 The data parameter  raises even more questions. 

When i include the data parameter, which points to the .yaml file (I use it for model traning),  I get the following error:

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[71], line 3
      1 from ultralytics.data.dataset import YOLODataset
----> 3 YOLODataset(data='/Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/config.yaml')

```
File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/dataset.py:41, in YOLODataset.__init__(self, data, use_segments, use_keypoints, *args, **kwargs)
     39 self.data = data
     40 assert not (self.use_segments and self.use_keypoints), 'Can not use both segments and keypoints.'
---> 41 super().__init__(*args, **kwargs)

TypeError: BaseDataset.__init__() missing 1 required positional argument: 'img_path'
```

It seems that it still demands the path to the folder with images. 

When I include  the image path, I get the following error:

```
AttributeError                            Traceback (most recent call last)
Cell In[73], line 3
      1 from ultralytics.data.dataset import YOLODataset
----> 3 YOLODataset(data='/Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/config.yaml',img_path='/Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/arctic/data/train/images')

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/dataset.py:41, in YOLODataset.__init__(self, data, use_segments, use_keypoints, *args, **kwargs)
     39 self.data = data
     40 assert not (self.use_segments and self.use_keypoints), 'Can not use both segments and keypoints.'
---> 41 super().__init__(*args, **kwargs)

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/base.py:97, in BaseDataset.__init__(self, img_path, imgsz, cache, augment, hyp, prefix, rect, batch_size, stride, pad, single_cls, classes, fraction)
     94     self.cache_images(cache)
     96 # Transforms
---> 97 self.transforms = self.build_transforms(hyp=hyp)

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/dataset.py:142, in YOLODataset.build_transforms(self, hyp)
    140     hyp.mosaic = hyp.mosaic if self.augment and not self.rect else 0.0
    141     hyp.mixup = hyp.mixup if self.augment and not self.rect else 0.0
--> 142     transforms = v8_transforms(self, self.imgsz, hyp)
    143 else:
    144     transforms = Compose([LetterBox(new_shape=(self.imgsz, self.imgsz), scaleup=False)])

File ~/anaconda3/lib/python3.11/site-packages/ultralytics/data/augment.py:931, in v8_transforms(dataset, imgsz, hyp, stretch)
    919 """Convert images to a size suitable for YOLOv8 training."""
    920 pre_transform = Compose([
    921     Mosaic(dataset, imgsz=imgsz, p=hyp.mosaic),
    922     CopyPaste(p=hyp.copy_paste),
   (...)
    929         pre_transform=None if stretch else LetterBox(new_shape=(imgsz, imgsz)),
    930     )])
--> 931 flip_idx = dataset.data.get('flip_idx', [])  # for keypoints augmentation
    932 if dataset.use_keypoints:
    933     kpt_shape = dataset.data.get('kpt_shape', None)

AttributeError: 'str' object has no attribute 'get'
```

**Reply by glenn-jocher** [reply]
@NS-Nik it seems like there's a misunderstanding with the `data` parameter. The `data` parameter should point to your `.yaml` file, and you don't need to specify `img_path` separately. Please ensure your `.yaml` file is correctly formatted and includes paths to your images and labels. If the issue persists, try updating to the latest version of the Ultralytics package.

**Comment by NS-Nik** [comment]
@glenn-jocher 
I guess that I am missing some specific settings in the .yaml file. 

Here is my code for YOLODataset initialization: 

`YOLODataset(data='/Users/user/Projects/MIZEddies/yolo/config.yaml'). `

Here is the structure of the config.yaml file: 

```
path: /Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/arctic/data
train: /Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/arctic/data/train
val: /Users/user/Yandex.Disk.localized/Projects/MIZEddies/yolo/arctic/data/valid
nc: 2

#classes

names:
  0: CE
  1: AE
```
I have the latest version of the Ultralytics package.

**Reply by glenn-jocher** [reply]
It looks like your `config.yaml` file is correctly structured. To visualize augmentations for a specific image, you can use the `augment` method from the `YOLO` class. Here's a minimal example:

```python
from ultralytics import YOLO
from PIL import Image

# Load a model
model = YOLO('yolov8n.pt')

# Load an image
img = Image.open('/path/to/your/image.jpg')

# Apply augmentations
augmented_images = model.augment(img)

# Display augmented images
for i, aug_img in enumerate(augmented_images):
    aug_img.show(title=f'Augmented Image {i}')
```

This will generate and display augmented versions of your specified image.

**Reply by NS-Nik** [reply]
@glenn-jocher 
This didn't work either.

The code returns an error:

` AttributeError: 'YOLO' object has no attribute 'augment`

I tried to work directly with albumentations library and for some augmentations techniques  its working. 

Here is the example of my code for hsv augmentation:

```
from PIL import Image
import albumentations as A
import numpy as np

img = Image.open('/Users/user/Desktop/image/ex2/original.jpg')
img = np.array(img)

#Define an augmentation pipeline for hsv
transform = A.Compose([
    A.HueSaturationValue(hue_shift_limit=50, sat_shift_limit=50, val_shift_limit=50, p=1)  
])
augmented_img = transform(image=img)['image']
augmented_img = Image.fromarray(augmented_img)

```

**Reply by glenn-jocher** [reply]
@NS-Nik the `YOLO` class does not have an `augment` method. Instead, you can use the `albumentations` library directly for augmentations, as you've done. Your current approach with `albumentations` is correct for applying HSV augmentation. If you need further assistance with YOLO-specific augmentations, please refer to the [Ultralytics documentation](https://docs.ultralytics.com/guides/preprocessing_annotated_data/).

**Reply by NS-Nik** [reply]
I have a felling that i've been talking to the AI bot:) 
Anyway, thanks for the help.

**Reply by glenn-jocher** [reply]
You're welcome! If you have any more questions or need further assistance, feel free to ask.
